snuba/datasets/factory.py

```diff
@@ -1,93 +1,105 @@
-from typing import Callable, MutableMapping, Sequence, Set
+from typing import Generator, MutableMapping, Sequence, Type
 
 from snuba import settings
 from snuba.datasets.dataset import Dataset
 from snuba.util import with_span
+from snuba.utils.config_component_factory import ConfigComponentFactory
 from snuba.utils.serializable_exception import SerializableException
 
-DATASETS_IMPL: MutableMapping[str, Dataset] = {}
-DATASETS_NAME_LOOKUP: MutableMapping[Dataset, str] = {}
-
-DEV_DATASET_NAMES: Set[str] = set()
-
-DATASET_NAMES: Set[str] = {
-    "discover",
-    "events",
-    "groupassignee",
-    "groupedmessage",
-    "metrics",
-    "outcomes",
-    "outcomes_raw",
-    "sessions",
-    "transactions",
-    "profiles",
-    "functions",
-    "generic_metrics",
-    "replays",
-    *(DEV_DATASET_NAMES if settings.ENABLE_DEV_FEATURES else set()),
-}
+
+class _DatasetFactory(ConfigComponentFactory[Dataset, str]):
+    def __init__(self) -> None:
+        self._dataset_map: MutableMapping[str, Dataset] = {}
+        self._name_map: MutableMapping[Type[Dataset], str] = {}
+        self.initialize()
+
+    def initialize(self) -> None:
```

Comment:
```suggestion
    def _initialize(self) -> None:
```
I would say this does not need to be a public method at this time, it can probably be removed from the ConfigComponentFactory as well or made private there.